src/untyped_array.rs

```diff
@@ -0,0 +1,254 @@
+//! Safe, untyped interface for NumPy's [N-dimensional arrays][ndarray]
+//!
+//! [ndarray]: https://numpy.org/doc/stable/reference/arrays.ndarray.html
+use std::{os::raw::c_int, slice};
+
+use pyo3::{
+    ffi, pyobject_native_type_extract, pyobject_native_type_named, AsPyPointer, IntoPy, PyAny,
+    PyNativeType, PyObject, PyTypeInfo, Python,
+};
+
+use crate::cold;
+use crate::dtype::PyArrayDescr;
+use crate::npyffi;
+
+/// A safe, untyped wrapper for NumPy's [`ndarray`][ndarray] class.
+///
+/// Unlike [`PyArray<T,D>`][crate::PyArray], this type does not constrain either element type `T` nor the dimensionality `D`.
+/// This can be useful to inspect function arguments, but it prevents operating on the elements without further downcasts.
+///
+/// # Example
+///
+/// Taking `PyUntypedArray` can be helpful to implement polymorphic entry points:
+///
+/// ```
+/// # use pyo3::prelude::*;
+/// use pyo3::exceptions::PyTypeError;
+/// use numpy::{Element, PyUntypedArray, PyArray1, dtype};
+///
+/// #[pyfunction]
+/// fn entry_point(py: Python, array: &PyUntypedArray) -> PyResult<()> {
+///     fn implementation<T: Element>(array: &PyUntypedArray) -> PyResult<()> {
+///         let array: &PyArray1<T> = array.downcast()?;
+///
+///         /* .. */
+///
+///         Ok(())
+///     }
+///
+///     let element_type = array.dtype();
+///
+///     if element_type.is_equiv_to(dtype::<f32>(py)) {
+///         implementation::<f32>(array)
```

Comment:
I feel like it would be a bit more practical to call `some_generic_function<T>` after downcasting the array to the concrete type. I'm expecting the case where users want to call some external genetic function, for example.

Reply:
I think both cases will happen and chose this one because it avoid repeating the downcast. That aside, I would move it outside to make it more prominent.